File: src/main/java/uk/edu/populationfitness/models/genes/sphere/ExponentialGenes.java
package uk.edu.populationfitness.models.genes.sphere;

import uk.edu.populationfitness.models.Config;
import uk.edu.populationfitness.models.fastmaths.ExpensiveCalculatedValues;
import uk.edu.populationfitness.models.fastmaths.ValueCalculator;
import uk.edu.populationfitness.models.genes.bitset.NormalizingBitSetGenes;

public class ExponentialGenes extends NormalizingBitSetGenes {
    private static class NormalizationRatioCalculator implements ValueCalculator<Double> {
        @Override
        public Double calculateValue(long n) {
            return 1.0 - Math.exp((long)(-0.5 * n));
        }
    }

    private static final ExpensiveCalculatedValues<Double> NormalizationRatios = new ExpensiveCalculatedValues(new ExponentialGenes.NormalizationRatioCalculator());

    public ExponentialGenes(Config config) {
        super(config, 1.0);
    }

    @Override
    protected double calculateNormalizationRatio(int n) {
        return NormalizationRatios.findOrCalculate(n);
    }

    @Override
    protected double calculateFitnessFromGenes(double[] unknowns) {
        /*
          f left (x right ) =- exp left (-0.5 sum from {i=2} to {n} {{x} rsub {i} rsup {2}} right )
         */

        double fitness = 0.0;
        final long scale = unknowns.length > 1000 ? unknowns.length / 1000 : 1;

        for (double x : unknowns) {
            double y = x / scale;

            fitness += y * y;
        }

        return 1.0 - Math.exp(-0.5 * fitness);
    }
}

File: DotNet/PopulationFitness/PopulationFitness/Models/Genes/Sphere/ExponentialGenes.cs
using PopulationFitness.Models.FastMaths;
using PopulationFitness.Models.Genes.BitSet;
using System;

namespace PopulationFitness.Models.Genes.Sphere
{
    public class ExponentialGenes : NormalizingBitSetGenes
    {
        private class NormalizationRatioCalculator : IValueCalculator<Double>
        {
            public Double CalculateValue(long n)
            {
                return 1.0 - Math.Exp(-0.5 * n);
            }
        }

        private static readonly ExpensiveCalculatedValues<double> NormalizationRatios = new ExpensiveCalculatedValues<double>(new ExponentialGenes.NormalizationRatioCalculator());

        public ExponentialGenes(Config config) : base(config, 1.0)
        {
        }

        protected override double CalculateNormalizationRatio(int n)
        {
            return NormalizationRatios.FindOrCalculate(n);
        }

        protected override double CalculateFitnessFromIntegers(long[] integer_values)
        {
            /*
              f left (x right ) =- exp left (-0.5 sum from {i=2} to {n} {{x} rsub {i} rsup {2}} right )
             */

            double fitness = 0.0;

            foreach (long integer_value in integer_values)
            {
                double x = Interpolate(integer_value);

                fitness += x * x;
            }

            return 1.0 - Math.Exp(-0.5 * fitness);
        }
    }
}
